core-rust/node-common/src/scheduler.rs

```diff
@@ -62,190 +62,216 @@
+            return running_task;
+        };
+        // The `if` below seems larger than required - this is for "incompatible types" reasons
+        if let Some(metrics) = self.metrics {
+            let task = metrics.wrap(task, self.name.as_str());
+            Self::spawn_periodic_future(runtime, interval, task, cancellation_token)
+        } else {
+            Self::spawn_periodic_future(runtime, interval, task, cancellation_token)
+        };
+        if let Some(task_tracker) = self.running_task_tracker {
+            task_tracker.track(running_task.clone());
+        }
+        running_task
     }
-}
-
-impl Scheduler for TokioScheduler {
-    type CancellationTokenType = CancellationToken;
 
-    fn start_periodic(
-        &self,
-        duration: Duration,
+    fn spawn_periodic_future(
+        runtime: &Runtime,
+        interval: Duration,
         mut task: impl 'static + FnMut() + Send,
-    ) -> Self::CancellationTokenType {
-        let cancellation_token = Self::CancellationTokenType::new();
-        let cloned_token = cancellation_token.clone();
-        self.runtime.spawn(async move {
-            let mut interval = tokio::time::interval(duration);
-
+        cancellation_token: CancellationToken,
+    ) {
+        runtime.spawn(async move {
+            let mut interval = tokio::time::interval(interval);
             loop {
                 tokio::select! {
-                    _ = cloned_token.cancelled() => {
+                    _ = cancellation_token.cancelled() => {
                         break;
                     },
                     _ = interval.tick() => {
-                       task();
+                        task();
                     },
                 }
             }
         });
-        cancellation_token
     }
 }
 
-/// An auxiliary structure to be used along side a [`Scheduler`] in order to collect
-/// [`CancellableToken`]s for the spawned tasks in order to autostop them on drop.
-pub struct TaskTracker<CT: CancellableToken> {
-    tasks: Vec<CancelOnDrop<CT>>,
+/// A handle representing a running task started using the scheduler.
+/// Dropping *all* its clones cancels any subsequent executions of the task.
+#[derive(Clone)]
+pub struct RunningTask {
+    cancellation_token: Arc<CancellationToken>,
 }
 
-impl<CT: CancellableToken> TaskTracker<CT> {
-    pub fn new() -> Self {
-        Self { tasks: Vec::new() }
-    }
-
-    pub fn track(&mut self, cancel_token: CT) {
-        self.tasks.push(CancelOnDrop::new(cancel_token));
+impl RunningTask {
+    fn new(cancellation_token: CancellationToken) -> Self {
+        Self {
+            cancellation_token: Arc::new(cancellation_token),
+        }
     }
 }
 
-impl<CT: CancellableToken> Default for TaskTracker<CT> {
-    fn default() -> Self {
-        Self::new()
+impl Drop for RunningTask {
+    fn drop(&mut self) {
+        if let Some(cancellation_token) = Arc::get_mut(&mut self.cancellation_token) {
```

Comment:
no, the `Arc::get_mut` only returns `Some` when it is _the only_ `Arc<>`. Since we are inside `drop()`, we additionally know that we have _the last_ `Arc<>`. So we cancel it and then the `Arc<>` is dropped, and its content (`CancellationToken`) gets dropped too and no-one cancels it anymore.

(AFAIU, the tokio async future holds a _clone_ of the `CancellationToken` contained here, but that's expected, because it wants to notice the cancelation that happens here, right?)

For context - my goal here is:
I want to hand out a couple of `RunningTask` handles (namely: one for the `TaskTracker` [if configured], one for the `.start_periodic()` caller). And the caller can also clone it if they have a reason.
And then, I want to cancel the task _not_ when the first of them is dropped, but after _all_ of them are dropped. That's it.

Reply:
Right! Thanks for explanation :D this is cool!